File: examples/05-chat-server/cpp/chat-server.cpp
/**
 * Chat Server Actor Pattern - 실시간 채팅 서버 예제 in C++
 * CAF 스타일로 구현
 *
 * 컴파일: g++ -std=c++17 chat-server.cpp -o chat-server -pthread
 */

#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <memory>
#include <chrono>
#include <thread>
#include <mutex>
#include <any>
#include <sstream>

// ============================================
// 타입 정의
// ============================================

enum class UserStatus { Online, Away, Offline };

std::string statusToString(UserStatus status) {
    switch (status) {
        case UserStatus::Online: return "online";
        case UserStatus::Away: return "away";
        case UserStatus::Offline: return "offline";
        default: return "unknown";
    }
}

struct UserInfo {
    std::string userId;
    std::string username;
    UserStatus status;
    std::chrono::system_clock::time_point joinedAt;
};

struct ChatMessage {
    std::string id;
    std::string senderId;
    std::string senderName;
    std::string content;
    std::chrono::system_clock::time_point timestamp;
    std::string type; // "public", "private", "system"
};

// ============================================
// 메시지 타입
// ============================================

struct JoinRoom {
    std::string userId;
    std::string username;
};

struct LeaveRoom {
    std::string userId;
};

struct SendMessage {
    std::string userId;
    std::string content;
};

struct DirectMessage {
    std::string fromUserId;
    std::string toUserId;
    std::string content;
};

struct SetStatus {
    std::string userId;
    UserStatus status;
};

// User 세션 메시지
struct MessageReceived {
    ChatMessage message;
};

struct UserJoined {
    UserInfo user;
};

struct UserLeft {
    std::string userId;
    std::string username;
};

struct UserStatusChanged {
    std::string userId;
    UserStatus status;
};

struct RoomHistory {
    std::vector<ChatMessage> messages;
};

// ============================================
// UserSession Actor
// ============================================

class UserSessionActor {
public:
    UserSessionActor(const std::string& userId, const std::string& username)
        : sessionId_(userId), username_(username) {}

    void receive(const std::any& msg) {
        if (msg.type() == typeid(MessageReceived)) {
            auto m = std::any_cast<MessageReceived>(msg);
            messages_.push_back(m.message);
            std::string prefix = m.message.type == "private" ? "[DM] " : "";
            std::cout << "  [" << username_ << "] " << prefix
                      << m.message.senderName << ": " << m.message.content << "\n";
        }
        else if (msg.type() == typeid(UserJoined)) {
            auto m = std::any_cast<UserJoined>(msg);
            std::cout << "  [" << username_ << "] " << m.user.username << " joined\n";
        }
        else if (msg.type() == typeid(UserLeft)) {
            auto m = std::any_cast<UserLeft>(msg);
            std::cout << "  [" << username_ << "] " << m.username << " left\n";
        }
        else if (msg.type() == typeid(UserStatusChanged)) {
            auto m = std::any_cast<UserStatusChanged>(msg);
            std::cout << "  [" << username_ << "] User " << m.userId
                      << " is now " << statusToString(m.status) << "\n";
        }
        else if (msg.type() == typeid(RoomHistory)) {
            auto m = std::any_cast<RoomHistory>(msg);
            std::cout << "  [" << username_ << "] Received "
                      << m.messages.size() << " messages from history\n";
        }
    }

private:
    std::string sessionId_;
    std::string username_;
    std::vector<ChatMessage> messages_;
};

// ============================================
// ChatRoom Actor
// ============================================

class ChatRoomActor {
public:
    ChatRoomActor(const std::string& roomId, const std::string& name)
        : roomId_(roomId), name_(name), maxHistorySize_(100), msgCounter_(0) {}

    void receive(const std::any& msg) {
        std::lock_guard<std::mutex> lock(mutex_);

        if (msg.type() == typeid(JoinRoom)) {
            handleJoinRoom(std::any_cast<JoinRoom>(msg));
        }
        else if (msg.type() == typeid(LeaveRoom)) {
            handleLeaveRoom(std::any_cast<LeaveRoom>(msg));
        }
        else if (msg.type() == typeid(SendMessage)) {
            handleSendMessage(std::any_cast<SendMessage>(msg));
        }
        else if (msg.type() == typeid(DirectMessage)) {
            handleDirectMessage(std::any_cast<DirectMessage>(msg));
        }
        else if (msg.type() == typeid(SetStatus)) {
            handleSetStatus(std::any_cast<SetStatus>(msg));
        }
    }

private:
    void handleJoinRoom(const JoinRoom& m) {
        if (users_.find(m.userId) != users_.end()) {
            std::cout << "[Room " << name_ << "] " << m.username << " is already in the room\n";
            return;
        }

        UserInfo user{m.userId, m.username, UserStatus::Online, std::chrono::system_clock::now()};
        users_[m.userId] = user;

        auto session = std::make_shared<UserSessionActor>(m.userId, m.username);
        userSessions_[m.userId] = session;

        std::cout << "[Room " << name_ << "] " << m.username
                  << " joined (" << users_.size() << " users)\n";

        // 시스템 메시지
        auto sysMsg = createSystemMessage(m.username + " joined the room");
        addToHistory(sysMsg);

        // 다른 사용자들에게 알림
        broadcastToOthers(m.userId, UserJoined{user});

        // 새 사용자에게 이력 전송
        std::vector<ChatMessage> history;
        size_t start = messageHistory_.size() > 20 ? messageHistory_.size() - 20 : 0;
        for (size_t i = start; i < messageHistory_.size(); i++) {
            history.push_back(messageHistory_[i]);
        }
        session->receive(RoomHistory{history});
    }

    void handleLeaveRoom(const LeaveRoom& m) {
        auto it = users_.find(m.userId);
        if (it == users_.end()) return;

        std::string username = it->second.username;
        users_.erase(it);
        userSessions_.erase(m.userId);

        std::cout << "[Room " << name_ << "] " << username
                  << " left (" << users_.size() << " users)\n";

        auto sysMsg = createSystemMessage(username + " left the room");
        addToHistory(sysMsg);

        broadcastToAll(UserLeft{m.userId, username});
    }

    void handleSendMessage(const SendMessage& m) {
        auto it = users_.find(m.userId);
        if (it == users_.end()) return;

        ChatMessage chatMsg{
            generateId(),
            m.userId,
            it->second.username,
            m.content,
            std::chrono::system_clock::now(),
            "public"
        };

        addToHistory(chatMsg);
        std::cout << "[Room " << name_ << "] " << it->second.username << ": " << m.content << "\n";

        broadcastToAll(MessageReceived{chatMsg});
    }

    void handleDirectMessage(const DirectMessage& m) {
        auto fromIt = users_.find(m.fromUserId);
        auto toSessionIt = userSessions_.find(m.toUserId);
        auto toUserIt = users_.find(m.toUserId);

        if (fromIt == users_.end() || toSessionIt == userSessions_.end() || toUserIt == users_.end()) {
            std::cout << "[Room " << name_ << "] DM failed: user not found\n";
            return;
        }

        ChatMessage dmMsg{
            generateId(),
            m.fromUserId,
            fromIt->second.username,
            m.content,
            std::chrono::system_clock::now(),
            "private"
        };

        std::cout << "[Room " << name_ << "] DM from " << fromIt->second.username
                  << " to " << toUserIt->second.username << ": " << m.content << "\n";

        // 수신자에게 전송
        toSessionIt->second->receive(MessageReceived{dmMsg});

        // 발신자에게도 전송
        auto fromSessionIt = userSessions_.find(m.fromUserId);
        if (fromSessionIt != userSessions_.end()) {
            fromSessionIt->second->receive(MessageReceived{dmMsg});
        }
    }

    void handleSetStatus(const SetStatus& m) {
        auto it = users_.find(m.userId);
        if (it == users_.end()) return;

        it->second.status = m.status;
        std::cout << "[Room " << name_ << "] " << it->second.username
                  << " is now " << statusToString(m.status) << "\n";

        broadcastToOthers(m.userId, UserStatusChanged{m.userId, m.status});
    }

    ChatMessage createSystemMessage(const std::string& content) {
        return ChatMessage{
            generateId(),
            "system",
            "System",
            content,
            std::chrono::system_clock::now(),
            "system"
        };
    }

    void addToHistory(const ChatMessage& msg) {
        messageHistory_.push_back(msg);
        if (messageHistory_.size() > maxHistorySize_) {
            messageHistory_.erase(messageHistory_.begin());
        }
    }

    template<typename T>
    void broadcastToAll(const T& msg) {
        for (auto& [_, session] : userSessions_) {
            session->receive(msg);
        }
    }

    template<typename T>
    void broadcastToOthers(const std::string& excludeId, const T& msg) {
        for (auto& [id, session] : userSessions_) {
            if (id != excludeId) {
                session->receive(msg);
            }
        }
    }

    std::string generateId() {
        msgCounter_++;
        auto now = std::chrono::system_clock::now();
        auto timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
            now.time_since_epoch()).count();
        std::stringstream ss;
        ss << timestamp << "-" << msgCounter_;
        return ss.str();
    }

    std::string roomId_;
    std::string name_;
    std::map<std::string, UserInfo> users_;
    std::map<std::string, std::shared_ptr<UserSessionActor>> userSessions_;
    std::vector<ChatMessage> messageHistory_;
    size_t maxHistorySize_;
    long long msgCounter_;
    std::mutex mutex_;
};

// ============================================
// 메인 함수
// ============================================

int main() {
    std::cout << "=== Chat Server Actor Pattern Demo (C++) ===\n\n";

    ChatRoomActor generalRoom("general", "General");

    std::cout << "--- 1. Users Joining ---\n\n";

    generalRoom.receive(JoinRoom{"alice", "Alice"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(JoinRoom{"bob", "Bob"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(JoinRoom{"charlie", "Charlie"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n--- 2. Public Messages ---\n\n";

    generalRoom.receive(SendMessage{"alice", "Hello everyone!"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(SendMessage{"bob", "Hi Alice!"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(SendMessage{"charlie", "Hey all!"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n--- 3. Direct Messages ---\n\n";

    generalRoom.receive(DirectMessage{"alice", "bob", "Hey Bob, can we talk privately?"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(DirectMessage{"bob", "alice", "Sure, what is it?"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n--- 4. Status Changes ---\n\n";

    generalRoom.receive(SetStatus{"charlie", UserStatus::Away});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n--- 5. More Messages ---\n\n";

    generalRoom.receive(SendMessage{"alice", "Is Charlie still here?"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(SetStatus{"charlie", UserStatus::Online});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(SendMessage{"charlie", "I'm back!"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n--- 6. User Leaving ---\n\n";

    generalRoom.receive(LeaveRoom{"bob"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    generalRoom.receive(SendMessage{"alice", "Bye Bob!"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n=== Demo Complete ===\n";

    return 0;
}

File: examples/05-chat-server/go/chat-server.go
// Chat Server Actor Pattern - 실시간 채팅 서버 예제 in Go
// Proto.Actor 스타일로 구현

package main

import (
	"fmt"
	"sync"
	"time"
)

// ============================================
// 타입 정의
// ============================================

type UserStatus string

const (
	StatusOnline  UserStatus = "online"
	StatusAway    UserStatus = "away"
	StatusOffline UserStatus = "offline"
)

type UserInfo struct {
	UserId   string
	Username string
	Status   UserStatus
	JoinedAt time.Time
}

type ChatMessage struct {
	Id        string
	SenderId  string
	SenderName string
	Content   string
	Timestamp time.Time
	Type      string // "public", "private", "system"
}

// ============================================
// 메시지 타입
// ============================================

type JoinRoom struct {
	UserId   string
	Username string
}

type LeaveRoom struct {
	UserId string
}

type SendMessage struct {
	UserId  string
	Content string
}

type DirectMessage struct {
	FromUserId string
	ToUserId   string
	Content    string
}

type SetStatus struct {
	UserId string
	Status UserStatus
}

// User 세션 메시지
type MessageReceived struct {
	Message ChatMessage
}

type UserJoined struct {
	User UserInfo
}

type UserLeft struct {
	UserId   string
	Username string
}

type UserStatusChanged struct {
	UserId string
	Status UserStatus
}

type RoomHistory struct {
	Messages []ChatMessage
}

// ============================================
// UserSession Actor
// ============================================

type UserSessionActor struct {
	sessionId string
	username  string
	messages  []ChatMessage
}

func NewUserSessionActor(userId, username string) *UserSessionActor {
	return &UserSessionActor{
		sessionId: userId,
		username:  username,
		messages:  make([]ChatMessage, 0),
	}
}

func (u *UserSessionActor) Receive(msg interface{}) {
	switch m := msg.(type) {
	case MessageReceived:
		u.messages = append(u.messages, m.Message)
		prefix := ""
		if m.Message.Type == "private" {
			prefix = "[DM] "
		}
		fmt.Printf("  [%s] %s%s: %s\n", u.username, prefix, m.Message.SenderName, m.Message.Content)

	case UserJoined:
		fmt.Printf("  [%s] %s joined\n", u.username, m.User.Username)

	case UserLeft:
		fmt.Printf("  [%s] %s left\n", u.username, m.Username)

	case UserStatusChanged:
		fmt.Printf("  [%s] User %s is now %s\n", u.username, m.UserId, m.Status)

	case RoomHistory:
		fmt.Printf("  [%s] Received %d messages from history\n", u.username, len(m.Messages))
	}
}

// ============================================
// ChatRoom Actor
// ============================================

type ChatRoomActor struct {
	roomId         string
	name           string
	users          map[string]*UserInfo
	userSessions   map[string]*UserSessionActor
	messageHistory []ChatMessage
	maxHistorySize int
	mu             sync.RWMutex
	msgCounter     int64
}

func NewChatRoomActor(roomId, name string) *ChatRoomActor {
	return &ChatRoomActor{
		roomId:         roomId,
		name:           name,
		users:          make(map[string]*UserInfo),
		userSessions:   make(map[string]*UserSessionActor),
		messageHistory: make([]ChatMessage, 0),
		maxHistorySize: 100,
	}
}

func (c *ChatRoomActor) Receive(msg interface{}) {
	c.mu.Lock()
	defer c.mu.Unlock()

	switch m := msg.(type) {
	case JoinRoom:
		c.handleJoinRoom(m)
	case LeaveRoom:
		c.handleLeaveRoom(m)
	case SendMessage:
		c.handleSendMessage(m)
	case DirectMessage:
		c.handleDirectMessage(m)
	case SetStatus:
		c.handleSetStatus(m)
	}
}

func (c *ChatRoomActor) handleJoinRoom(m JoinRoom) {
	if _, exists := c.users[m.UserId]; exists {
		fmt.Printf("[Room %s] %s is already in the room\n", c.name, m.Username)
		return
	}

	user := &UserInfo{
		UserId:   m.UserId,
		Username: m.Username,
		Status:   StatusOnline,
		JoinedAt: time.Now(),
	}
	c.users[m.UserId] = user

	session := NewUserSessionActor(m.UserId, m.Username)
	c.userSessions[m.UserId] = session

	fmt.Printf("[Room %s] %s joined (%d users)\n", c.name, m.Username, len(c.users))

	// 시스템 메시지
	sysMsg := c.createSystemMessage(fmt.Sprintf("%s joined the room", m.Username))
	c.addToHistory(sysMsg)

	// 다른 사용자들에게 알림
	c.broadcastToOthers(m.UserId, UserJoined{User: *user})

	// 새 사용자에게 이력 전송
	historyStart := 0
	if len(c.messageHistory) > 20 {
		historyStart = len(c.messageHistory) - 20
	}
	session.Receive(RoomHistory{Messages: c.messageHistory[historyStart:]})
}

func (c *ChatRoomActor) handleLeaveRoom(m LeaveRoom) {
	user, exists := c.users[m.UserId]
	if !exists {
		return
	}

	delete(c.users, m.UserId)
	delete(c.userSessions, m.UserId)

	fmt.Printf("[Room %s] %s left (%d users)\n", c.name, user.Username, len(c.users))

	sysMsg := c.createSystemMessage(fmt.Sprintf("%s left the room", user.Username))
	c.addToHistory(sysMsg)

	c.broadcastToAll(UserLeft{UserId: m.UserId, Username: user.Username})
}

func (c *ChatRoomActor) handleSendMessage(m SendMessage) {
	user, exists := c.users[m.UserId]
	if !exists {
		return
	}

	chatMsg := ChatMessage{
		Id:         c.generateId(),
		SenderId:   m.UserId,
		SenderName: user.Username,
		Content:    m.Content,
		Timestamp:  time.Now(),
		Type:       "public",
	}

	c.addToHistory(chatMsg)
	fmt.Printf("[Room %s] %s: %s\n", c.name, user.Username, m.Content)

	c.broadcastToAll(MessageReceived{Message: chatMsg})
}

func (c *ChatRoomActor) handleDirectMessage(m DirectMessage) {
	fromUser, fromExists := c.users[m.FromUserId]
	toSession, toExists := c.userSessions[m.ToUserId]
	toUser := c.users[m.ToUserId]

	if !fromExists || !toExists || toUser == nil {
		fmt.Printf("[Room %s] DM failed: user not found\n", c.name)
		return
	}

	dmMsg := ChatMessage{
		Id:         c.generateId(),
		SenderId:   m.FromUserId,
		SenderName: fromUser.Username,
		Content:    m.Content,
		Timestamp:  time.Now(),
		Type:       "private",
	}

	fmt.Printf("[Room %s] DM from %s to %s: %s\n", c.name, fromUser.Username, toUser.Username, m.Content)

	// 수신자에게 전송
	toSession.Receive(MessageReceived{Message: dmMsg})

	// 발신자에게도 전송
	if fromSession, ok := c.userSessions[m.FromUserId]; ok {
		fromSession.Receive(MessageReceived{Message: dmMsg})
	}
}

func (c *ChatRoomActor) handleSetStatus(m SetStatus) {
	user, exists := c.users[m.UserId]
	if !exists {
		return
	}

	user.Status = m.Status
	fmt.Printf("[Room %s] %s is now %s\n", c.name, user.Username, m.Status)

	c.broadcastToOthers(m.UserId, UserStatusChanged{UserId: m.UserId, Status: m.Status})
}

func (c *ChatRoomActor) createSystemMessage(content string) ChatMessage {
	return ChatMessage{
		Id:         c.generateId(),
		SenderId:   "system",
		SenderName: "System",
		Content:    content,
		Timestamp:  time.Now(),
		Type:       "system",
	}
}

func (c *ChatRoomActor) addToHistory(msg ChatMessage) {
	c.messageHistory = append(c.messageHistory, msg)
	if len(c.messageHistory) > c.maxHistorySize {
		c.messageHistory = c.messageHistory[1:]
	}
}

func (c *ChatRoomActor) broadcastToAll(msg interface{}) {
	for _, session := range c.userSessions {
		session.Receive(msg)
	}
}

func (c *ChatRoomActor) broadcastToOthers(excludeId string, msg interface{}) {
	for id, session := range c.userSessions {
		if id != excludeId {
			session.Receive(msg)
		}
	}
}

func (c *ChatRoomActor) generateId() string {
	c.msgCounter++
	return fmt.Sprintf("%d-%d", time.Now().UnixNano(), c.msgCounter)
}

// ============================================
// 메인 함수
// ============================================

func main() {
	fmt.Println("=== Chat Server Actor Pattern Demo (Go) ===\n")

	generalRoom := NewChatRoomActor("general", "General")

	fmt.Println("--- 1. Users Joining ---\n")

	generalRoom.Receive(JoinRoom{UserId: "alice", Username: "Alice"})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(JoinRoom{UserId: "bob", Username: "Bob"})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(JoinRoom{UserId: "charlie", Username: "Charlie"})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n--- 2. Public Messages ---\n")

	generalRoom.Receive(SendMessage{UserId: "alice", Content: "Hello everyone!"})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(SendMessage{UserId: "bob", Content: "Hi Alice!"})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(SendMessage{UserId: "charlie", Content: "Hey all!"})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n--- 3. Direct Messages ---\n")

	generalRoom.Receive(DirectMessage{
		FromUserId: "alice",
		ToUserId:   "bob",
		Content:    "Hey Bob, can we talk privately?",
	})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(DirectMessage{
		FromUserId: "bob",
		ToUserId:   "alice",
		Content:    "Sure, what is it?",
	})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n--- 4. Status Changes ---\n")

	generalRoom.Receive(SetStatus{UserId: "charlie", Status: StatusAway})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n--- 5. More Messages ---\n")

	generalRoom.Receive(SendMessage{UserId: "alice", Content: "Is Charlie still here?"})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(SetStatus{UserId: "charlie", Status: StatusOnline})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(SendMessage{UserId: "charlie", Content: "I'm back!"})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n--- 6. User Leaving ---\n")

	generalRoom.Receive(LeaveRoom{UserId: "bob"})
	time.Sleep(100 * time.Millisecond)

	generalRoom.Receive(SendMessage{UserId: "alice", Content: "Bye Bob!"})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n=== Demo Complete ===")
}
